src/overlays/actors/ovl_En_Clear_Tag/z_en_clear_tag.c

```diff
@@ -61,25 +76,999 @@ static ColliderCylinderInit D_809D5C40 = {
+
+            effect->timer = effect->bounces = 0;
+
+            effect->floorHeight = floorHeight;
+
+            seed = (s32)Rand_ZeroFloat(10.0f);
+            effect->random = seed;
+
+            return;
+        }
+    }
+}
+
+/**
+ * Creates a fire effect.
+ * Fire effects are spawned by debris effects. Fire effects spawn smoke effects
+ */
+void EnClearTag_CreateFireEffect(GlobalContext* globalCtx, Vec3f* pos, f32 scale) {
+    s16 i;
+    s16 seed;
+    EnClearTagEffect* effect;
+    GlobalContext* globalCtx2;
+
+    globalCtx2 = globalCtx;
+    effect = (EnClearTagEffect*)globalCtx2->specialEffects;
+
+    // Look for an available effect to allocate a fire effect to.
+    for (i = 0; i < CLEAR_TAG_EFFECT_MAX_COUNT; i++, effect++) {
+        if (effect->type == CLEAR_TAG_EFFECT_AVAILABLE) {
+            seed = (s32)Rand_ZeroFloat(100.0f);
+            effect->random = seed;
+            effect->type = CLEAR_TAG_EFFECT_FIRE;
+
+            effect->position = *pos;
+            effect->velocity = sZeroVector;
+            effect->acceleration = sZeroVector;
+            effect->acceleration.y = 0.15f;
+
+            effect->scale = scale;
+
+            effect->primColor.a = 200.0f;
+
+            return;
+        }
+    }
+}
+
+/**
+ * Creates a smoke effect.
+ * Smoke effects are spawned by fire effects.
+ */
+void EnClearTag_CreateSmokeEffect(GlobalContext* globalCtx, Vec3f* position, f32 scale) {
+    s16 i;
+    s16 seed;
+    EnClearTagEffect* effect;
+    GlobalContext* globalCtx2;
+
+    globalCtx2 = globalCtx;
+    effect = (EnClearTagEffect*)globalCtx2->specialEffects;
+
+    // Look for an available effect to allocate a smoke effect to.
+    for (i = 0; i < CLEAR_TAG_EFFECT_MAX_COUNT; i++, effect++) {
+        if (effect->type == CLEAR_TAG_EFFECT_AVAILABLE) {
+            seed = (s32)Rand_ZeroFloat(100.0f);
+            effect->random = seed;
+            effect->type = CLEAR_TAG_EFFECT_SMOKE;
+
+            effect->position = *position;
+            effect->velocity = sZeroVector;
+            effect->acceleration = sZeroVector;
+
+            effect->scale = scale;
+            effect->maxScale = scale + scale;
+
+            effect->primColor.r = 200.0f;
+            effect->primColor.g = 20.0f;
+            effect->primColor.b = 0.0f;
+            effect->primColor.a = 255.0f;
+            effect->envColor.r = 255.0f;
+            effect->envColor.g = 215.0f;
+            effect->envColor.b = 255.0f;
+
+            return;
+        }
+    }
+}
+
+/**
+ * Creates a flash effect.
+ * Flash effects are spawned when the Arwing dies.
+ * Flash effects two components: 1) a billboard flash, and 2) a light effect on the ground.
+ */
+void EnClearTag_CreateFlashEffect(GlobalContext* globalCtx, Vec3f* position, f32 scale, f32 floorHeight,
+                                  Vec3f* floorTangent) {
+    s16 i;
+    EnClearTagEffect* effect;
+    GlobalContext* globalCtx2;
+
+    globalCtx2 = globalCtx;
```

Comment:
This recast should be unnecessary. Does it match without it?

Reply:
It's generally good practice to explicitly cast your void* pointers to the explicit type for readability, is it not?

Reply:
but this isn't a void* pointer.

Reply:
Do all of the special effects have a shared base struct?

Reply:
No I mean `globalCtx2` should be unnecessary.

Reply:
Yep. It looks like I copy/pasted the boiler plate code for this function. Only the engine functions _should_ need the base game struct.

Good catch.